lib/dangermattic/plugins/podfile_checks.rb

```diff
@@ -32,37 +32,85 @@ class PodfileChecks < Plugin
     #
     # @return [void]
     def check_podfile_does_not_have_commit_references(podfile_lock_path: DEFAULT_PODFILE_LOCK_PATH)
+      check_podfile_does_not_match(
+        regexp: COMMIT_REFERENCE_REGEXP,
+        podfile_lock_path: podfile_lock_path,
+        match_found_message_generator: ->(matches) { "Podfile reference(s) to a commit hash:\n```#{matches.join("\n")}```" }
+      )
+    end
+
+    # Check for Podfile references to commit hashes in the Podfile.lock in a pull request.
+    #
+    # @return [void]
+    def check_podfile_diff_does_not_have_commit_references
+      warning_message = 'This PR adds a Podfile reference to a commit hash:'
+      check_podfile_diff_entries_do_not_match(regexp: COMMIT_REFERENCE_REGEXP, match_found_message: warning_message)
+    end
+
+    # Check if the Podfile.lock contains any references to branches and raise a failure if it does.
+    #
+    # @param podfile_lock_path [String] (optional) The path to the Podfile.lock file.
+    #        Defaults to the `DEFAULT_PODFILE_LOCK_PATH` constant if not provided.
+    #
+    # @example Checking the default Podfile.lock:
+    #   check_podfile_does_not_have_branch_references
+    #
+    # @example Checking a custom Podfile.lock at a specific path:
+    #    check_podfile_does_not_have_branch_references(podfile_lock_path: '/path/to/Podfile.lock')
+    #
+    # @return [void]
+    def check_podfile_does_not_have_branch_references(podfile_lock_path: DEFAULT_PODFILE_LOCK_PATH)
```

Comment:
This was already present in the previous implementation and, to be honest, I cannot recall what was the precise original idea, but the full `Podfile` check reports an error and the `diff` check reports a warning. Probably just a verbatim of Peril's current behaviour.

I have a task for myself to review all plugins regarding error/warning reporting, perhaps adding the option for either option in more of them, or making some of them more strictly reporting errors, but perhaps, while at it, we can already give this case some thought.

What's your take for this plugin (and in general)?

If we want to give users the option to choose between error / warning with a default value, it should be rather simple (`git_utils.check_added_diff_lines()` already provides a `fail_on_error: false` parameter).

Reply:
Good question. I'll admit I didn't think through the design too much while working on this PR as I was motivated by getting it done ASAP—which is ironic in hindsight given I let it go stale 😳 

I think we have two questions here, right?

- Should the `Podfile` (and, generalizing, any lockfile) check inspect the entire file or the PR diff only?
- Should the check report a warning or an error?

Plus, the design question of whether to add a `fail_on_error:` configuration for the caller.

I'll start with the easiest one for me to answer. 

#### `fail_on_error`?

I'm 100% on board with adding a `fail_on_error:` param with a reasonable default to this and every other check. I think it aligns with the guiding principle of building _flexible tools with opinionated golden paths_.

In writing this I briefly considered suggesting even more flexibility, e.g. (`on_violation: :fail # :warn, :comment`). That would be neat, but I think that most of our checks work to provide the user feedback in terms of warnings or straight out failures, with the comment option used by only a few checks. Example from Peril:

<img width="927" alt="image" src="https://github.com/Automattic/dangermattic/assets/1218433/c7b22a3d-d569-420e-aea7-a3eacf8072ae">

Under that assumption, it seems that passing and handling a `Bool` would be much simpler that establishing and maintaining a consistent `:fail, :warn, :comment` pattern.

The only caveat I have is that I'd like to use the word "violation" instead of "error". That is `fail_on_violation: true` over `fail_on_error: true`. I feel like "error" is too harsh for the kind of check failures that Danger can pick up. _I also acknowledge this is overthinking it and it's easier to update. However, given we have many clients I think it's worth spending extra time to settle on something less likely to need changes in the future._

What do you think?

### Check the full lockfile or the diff only?

The case for checking the diff only is that Danger is a tool for "unit testing pull requests" so checking the diff seems like the most appropriate choice, given that PRs modify files in diffs.

Based on that rationale, I'd say to keep only the diff check. Even though intuitively I would expect Danger to check the whole file in this instance.

But if we generalize the diff vs whole file behavior, then it seems clear to me that diff is the way to go. Let's consider a linter like RuboCop. It would be rather noisy to get warnings about pre-existing violations that a PR did not introduce, despite the fact that those violations ought to be addressed ASAP.

So, for the sake of consistency, I think diff-only is the approach we should take here, too.

### Warn or fail when user fetches pods by commit?

The operating procedure we agree upon, paNNhX-K4-p2, allow references to commits to be used during development, so I think a warning to raise awareness is enough here.

...which just made me realize that the check for branch references which I considered useless and kept just for reference actually becomes more useful that the check for commit. Because, while using commits is okay except before making a release build (which we could implement by changing the `fail_on_violation` value depending on the base branch) referencing branches should never be allowed, we can therefore add that check, too to our `Dangerfile` with `fail_on_violation: true`.

Reply:
> The only caveat I have is that I'd like to use the word "violation" instead of "error". That is `fail_on_violation: true` over `fail_on_error: true`. I feel like "error" is too harsh for the kind of check failures that Danger can pick up. _I also acknowledge this is overthinking it and it's easier to update. However, given we have many clients I think it's worth spending extra time to settle on something less likely to need changes in the future._
> 
> What do you think?

Agreed! Plus, when `fail_on_error` is `false`, it's not really an "error" if we're going to report a warning 😄 
A parameter with the options `:fail`, `:warn`, `:comment` also isn't a bad idea -- it's something [we implemented](https://github.com/Automattic/dangermattic/blob/4df9a409f6a5d78a63b55d6e5fb8bdb27a5efd8a/lib/dangermattic/plugins/milestone_checker.rb#L59) in another plugin a slightly different purpose.

What about calling it `severity:` or `report:` with the  options `:fail`, `:warn`, `:comment`?
I'll work in the coming days on a set of improvements for Dangermattic, and one of them will be on error reporting, so I can pick this up.

> ### Check the full lockfile or the diff only?
. [...]
> But if we generalize the diff vs whole file behavior, then it seems clear to me that diff is the way to go. Let's consider a linter like RuboCop. It would be rather noisy to get warnings about pre-existing violations that a PR did not introduce, despite the fact that those violations ought to be addressed ASAP.

That's also the thing: Danger RuboCop and Danger SwiftLint both offer the option to lint everything or lint only the diff, and perhaps this philosophy is worth incorporating as well, at least in some cases.

> ### Warn or fail when user fetches pods by commit?
> The operating procedure we agree upon, [paNNhX-K4-p2](https://href.li/?https://wp.me/paNNhX-K4), allow references to commits to be used during development, so I think a warning to raise awareness is enough here.

👍 

> ...which just made me realize that the check for branch references which I considered useless and kept just for reference actually becomes more useful that the check for commit. Because, while using commits is okay except before making a release build (which we could implement by changing the `fail_on_violation` value depending on the base branch) referencing branches should never be allowed, we can therefore add that check, too to our `Dangerfile` with `fail_on_violation: true`.

Sounds good 👍

Reply:
@iangmaia thank you for promptly following up—_not like I do_. 

Based on the discussion above, is there anything left to address in this PR, or can we merge it?

Reply:
Let's merge it! :shipit: